internal/forms/forms.go

```diff
@@ -237,6 +246,182 @@ func (m *Manager) GetFormTemplateHandler(w http.ResponseWriter, r *http.Request)
+	if err != nil {
+		http.Error(w, fmt.Sprint(err), http.StatusInternalServerError)
+		return
+	}
+	jsn, _ := json.Marshal(response)
+	_, _ = w.Write(jsn)
+}
+
+// UpdateFormTemplates handles searching for and installing the latest version of the form templates.
+func (m *Manager) UpdateFormTemplates() (UpdateResponse, error) {
+	log.Printf("Updating form templates; current version is %v", m.getFormsVersion())
+	newestVersion, downloadLink, err := m.getLatestFormsInfo()
+	if err != nil {
+		return UpdateResponse{}, err
+	}
+	if !m.isNewerVersion(newestVersion) {
+		log.Printf("Latest forms version is %v; nothing to do", newestVersion)
+		return UpdateResponse{
+			NewestVersion: newestVersion,
+			Action:        "none",
+		}, nil
+	}
+
+	err = m.downloadAndUnzipForms(downloadLink)
+	if err != nil {
+		return UpdateResponse{}, err
+	}
+	log.Printf("Finished forms update to %v", newestVersion)
+	// TODO: re-init forms manager
+	return UpdateResponse{
+		NewestVersion: newestVersion,
+		Action:        "update",
+	}, nil
+}
+
+func (m *Manager) getLatestFormsInfo() (string, string, error) {
+	client := http.Client{Timeout: 10 * time.Second}
+	req, _ := http.NewRequest("GET", formsVersionInfoUrl, nil)
+	req.Header.Set("User-Agent", m.config.UserAgent)
+	versionResp, err := client.Do(req)
+	if err != nil || versionResp.StatusCode != http.StatusOK {
+		err := fmt.Errorf("can't fetch winlink forms version page: %v", err)
+		log.Print(err)
+		return "", "", err
+	}
+	bodyBytes, err := ioutil.ReadAll(versionResp.Body)
+	if err != nil {
+		err := fmt.Errorf("can't read winlink forms version page: %v", err)
+		log.Print(err)
+		return "", "", err
+	}
+	bodyString := string(bodyBytes)
+
+	// Scrape for the version and download link
+	versionRe := regexp.MustCompile(`Standard_Forms - Version (\d+\.\d+\.\d+(\.\d+)?)`)
+	downloadRe := regexp.MustCompile(`https:\/\/1drv.ms\/u\/([a-zA-Z0-9-_!]+)\?e=([a-zA-Z0-9-_]+)`)
+	versionMatches := versionRe.FindStringSubmatch(bodyString)
+	downloadMatches := downloadRe.FindStringSubmatch(bodyString)
+	if versionMatches == nil || len(versionMatches) < 2 || downloadMatches == nil || len(downloadMatches) < 3 {
+		err := errors.New("can't scrape the version info page, HTML structure may have changed")
+		log.Print(err)
+		return "", "", err
+	}
+	newestVersion := versionMatches[1]
+	docId := downloadMatches[1]
+	auth := downloadMatches[2]
+	downloadLink := "https://api.onedrive.com/v1.0/shares/" + docId + "/root/content?e=" + auth
+	return newestVersion, downloadLink, nil
+}
+
+func (m *Manager) downloadAndUnzipForms(downloadLink string) error {
+	log.Printf("Updating forms via %v", downloadLink)
+	client := http.Client{Timeout: 30 * time.Second}
+	req, _ := http.NewRequest("GET", downloadLink, nil)
+	req.Header.Set("User-Agent", m.config.UserAgent)
+	downloadResp, err := client.Do(req)
+	if err != nil {
+		err := fmt.Errorf("can't download update ZIP: %v", err)
+		log.Print(err)
+		return err
```

Comment:
I'm getting rid of all of the `log.Print(err)`, and using `%w`, those makes sense. However, there's still logging like `log.Printf("Updating forms via %v", downloadLink)`; I don't think I can get rid of those easily unless I refactor to pass in a logger object. That's doable, but I'd prefer to do that in a separate PR.

Reply:
Yes, I think we can live with a couple of useful log output. Especially since this is an internal package 👍 We can improve it later if we feel like it.